language/move-core/types/src/language_storage.rs

```diff
@@ -157,3 +157,143 @@ impl Display for ResourceKey {
         write!(f, "0x{}/{}", self.address.short_str_lossless(), self.type_)
     }
 }
+
+// =================================================================================================
+// Type definition with free type variables
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub enum Type {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Signer,
+    Vector(Box<Type>),
+    Struct(StructType),
+    TypeArg(usize),
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct StructType {
+    pub address: AccountAddress,
+    pub module: Identifier,
+    pub name: Identifier,
+    pub type_params: Vec<Type>,
+}
+
+impl Type {
+    pub fn is_free(&self) -> bool {
+        match self {
+            Type::TypeArg(_) => true,
+            Type::Bool | Type::U8 | Type::U64 |  Type::U128 | Type::Address | Type::Signer | Type::Vector(_) => false,
+            Type::Struct(s) => s.is_free(),
+        }
+    }
+    pub fn subst(self, type_args: &[TypeTag]) -> Option<TypeTag> {
+        Some(match self {
+            Type::Bool => TypeTag::Bool,
+            Type::U8 => TypeTag::U8,
+            Type::U64 => TypeTag::U64,
+            Type::U128 => TypeTag::U128,
+            Type::Address => TypeTag::Address,
+            Type::Signer => TypeTag::Signer,
+            Type::Vector(ty) => TypeTag::Vector(Box::new(ty.subst(type_args)?)),
+            Type::Struct(s) => TypeTag::Struct(s.subst(type_args)?),
+            Type::TypeArg(idx) => type_args
+                .get(idx)?
+                .clone(),
+        })
+    }
+
+    pub fn to_type_tag(self) -> Option<TypeTag> {
+        self.subst(&[])
+    }
+}
+
+impl StructType {
+    pub fn is_free(&self) -> bool {
+        self.type_params.is_empty() || self.type_params.iter().all(|ty| ty.is_free())
```

Comment:
I am confused... If the struct is of a concrete type (i.e., no type params), then the struct type is free?